File: placement-flutter/storyly_placement_flutter/ios/Classes/Common/Data/Product/ProductData.swift
import Foundation
import StorylyPlacement
import StorylyCore

// MARK: - STRProductItem Encoding/Decoding

func encodeSTRProductItem(_ item: STRProductItem) -> [String: Any] {
    let result: [String: Any] = ([
        "productId": item.productId,
        "productGroupId": item.productGroupId,
        "title": item.title,
        "url": item.url,
        "desc": item.desc,
        "price": item.price,
        "currency": item.currency,
        "salesPrice": item.salesPrice,
        "lowestPrice": item.lowestPrice,
        "imageUrls": item.imageUrls,
        "ctaText": item.ctaText,
        "variants": item.variants?.map { encodeSTRProductVariant($0) }
    ] as [String: Any?]).compactMapValues { $0 }
    return result
}

func decodeSTRProductItem(_ dict: [String: Any]) -> STRProductItem? {
    guard let productId = dict["productId"] as? String,
          let productGroupId = dict["productGroupId"] as? String,
          let title = dict["title"] as? String,
          let url = dict["url"] as? String,
          let desc = dict["desc"] as? String,
          let currency = dict["currency"] as? String,
          let price = dict["price"] as? Double else {
        return nil
    }
    
    
    let salesPrice = dict["salesPrice"] as? Double
    let imageUrls = dict["imageUrls"] as? [String]
    let ctaText = dict["ctaText"] as? String
    let variants: [STRProductVariant]? = (dict["variants"] as? [[String: Any]])?.compactMap { decodeSTRProductVariant($0) }
    let lowestPrice = dict["lowestPrice"] as? NSNumber
    
    return STRProductItem(
        productId: productId,
        productGroupId: productGroupId,
        title: title,
        url: url,
        description: desc,
        price: Float(price),
        salesPrice: salesPrice as NSNumber?,
        lowestPrice: lowestPrice,
        currency: currency,
        imageUrls: imageUrls,
        variants: variants,
        ctaText: ctaText
    )
}

// MARK: - STRProductVariant Encoding/Decoding

func encodeSTRProductVariant(_ variant: STRProductVariant) -> [String: Any] {
    let result: [String: Any] = [
        "name": variant.name,
        "value": variant.value,
        "key": variant.key
    ]
    return result
}

func decodeSTRProductVariant(_ dict: [String: Any]) -> STRProductVariant? {
    guard let name = dict["name"] as? String,
          let value = dict["value"] as? String,
          let key = dict["key"] as? String else {
        return nil
    }
    return STRProductVariant(name: name, value: value, key: key)
}

// MARK: - STRCartItem Encoding/Decoding

func encodeSTRCartItem(_ item: STRCartItem?) -> [String: Any?]? {
    guard let item = item else { return nil }
    
    let result: [String: Any] = ([
        "product": encodeSTRProductItem(item.product),
        "quantity": item.quantity,
    ] as [String: Any?]).compactMapValues { $0 }
    return result
}

func decodeSTRCartItem(_ dict: [String: Any]) -> STRCartItem? {
    guard let productDict = dict["product"] as? [String: Any],
          let item = decodeSTRProductItem(productDict),
          let quantity = dict["quantity"] as? Int else {
        return nil
    }
    
    let totalPrice = dict["totalPrice"] as? Double
    let oldTotalPrice = dict["oldTotalPrice"] as? Double
    
    return STRCartItem(
        item: item,
        quantity: quantity
    )
}

// MARK: - STRProductInformation Encoding

func encodeSTRProductInformation(_ info: STRProductInformation) -> [String: Any?] {
    return [
        "productId": info.productId,
        "productGroupId": info.productGroupId
    ]
}

func decodeSTRProductInformation(_ dict: [String: Any]) -> STRProductInformation {
    let productId = dict["productId"] as? String
    let productGroupId = dict["productGroupId"] as? String
    return STRProductInformation(productId: productId, productGroupId: productGroupId)
}

File: placement-reactnative/storyly-placement-react-native/android/src/main/java/com/storylyplacementreactnative/common/data/product/ProductData.kt
package com.storylyplacementreactnative.common.data.product

import com.appsamurai.storyly.core.data.model.product.STRCart
import com.appsamurai.storyly.core.data.model.product.STRCartItem
import com.appsamurai.storyly.core.data.model.product.STRProductInformation
import com.appsamurai.storyly.core.data.model.product.STRProductItem
import com.appsamurai.storyly.core.data.model.product.STRProductVariant

internal fun encodeSTRProductItem(product: STRProductItem): Map<String, Any?> {
    return mapOf(
        "productId" to product.productId,
        "productGroupId" to product.productGroupId,
        "title" to product.title,
        "url" to product.url,
        "desc" to product.desc,
        "price" to product.price.toDouble(),
        "salesPrice" to product.salesPrice?.toDouble(),
        "lowestPrice" to product.lowestPrice?.toDouble(),
        "currency" to product.currency,
        "imageUrls" to product.imageUrls,
        "variants" to product.variants.map { variant -> encodeSTRProductVariant(variant) },
        "ctaText" to product.ctaText,
    )
}

internal fun decodeSTRProductItem(product: Map<String, Any?>?): STRProductItem? {
    product ?: return null
    val productId = product["productId"] as? String ?: return null
    val productGroupId = product["productGroupId"] as? String ?: return null
    val title = product["title"] as? String ?: return null
    val url = product["url"] as? String ?: return null
    val desc = product["desc"] as? String ?: return null
    val price = (product["price"] as? Number)?.toFloat() ?: return null
    return STRProductItem(
        productId = productId,
        productGroupId = productGroupId,
        title = title,
        url = url,
        desc = desc,
        price = price,
        salesPrice = (product["salesPrice"] as? Number)?.toFloat(),
        lowestPrice = (product["lowestPrice"] as? Number)?.toFloat(),
        currency = product["currency"] as String,
        imageUrls = product["imageUrls"] as? List<String>,
        variants = (product["variants"] as? List<Map<String, Any?>>)
            ?.mapNotNull { variant -> decodeSTRProductVariant(variant) } ?: emptyList(),
        ctaText = product["ctaText"] as? String,
    )
}

internal fun encodeSTRProductVariant(variant: STRProductVariant): Map<String, Any?> {
    return mapOf(
        "name" to variant.name,
        "value" to variant.value,
        "key" to variant.key,
    )
}

internal fun decodeSTRProductVariant(variant: Map<String, Any?>): STRProductVariant? {
    val name = variant["name"] as? String ?: return null
    val value = variant["value"] as? String ?: return null
    val key = variant["key"] as? String ?: return null
    return STRProductVariant(
        name = name,
        value = value,
        key = key,
    )
}

internal fun encodeSTRProductInformation(info: STRProductInformation): Map<String, Any?> {
    return mapOf(
        "productId" to info.productId,
        "productGroupId" to info.productGroupId,
    )
}

internal fun decodeSTRProductInformation(info: Map<String, Any?>): STRProductInformation? {
    val productId = info["productId"] as? String ?: return null
    val productGroupId = info["productGroupId"] as? String ?: return null
    return STRProductInformation(
        productId = productId,
        productGroupId = productGroupId,
    )
}

internal fun encodeSTRCartItem(item: STRCartItem?): Map<String, Any?>? {
    item ?: return null
    return mapOf(
        "product" to encodeSTRProductItem(item.product),
        "quantity" to item.quantity,
    )
}